azure/services/vmssextensions/spec.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vmssextensions
+
+import (
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-04-01/compute"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+)
+
+// VMSSExtensionSpec defines the specification for a VM or VMScaleSet extension.
+type VMSSExtensionSpec struct {
+	Name              string
+	ResourceGroup     string
+	VMName            string
+	Publisher         string
+	Version           string
+	ProtectedSettings map[string]string
+}
+
+// ResourceName returns the name of the VMSS extension.
+func (s *VMSSExtensionSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *VMSSExtensionSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName returns the name of the VMSS that owns this VMSS extension.
+func (s *VMSSExtensionSpec) OwnerResourceName() string {
+	return s.VMName
+}
+
+// Parameters returns the parameters for the VMSS extension.
+func (s *VMSSExtensionSpec) Parameters(existing interface{}) (interface{}, error) {
+	if existing != nil {
+		_, ok := existing.(compute.VirtualMachineScaleSetExtension)
+		if !ok {
+			return nil, errors.Errorf("%T is not a compute.VirtualMachineScaleSetExtension", existing)
+		}
+
+		// VMSS extension already exists, nothing to update.
+		return nil, nil
+	}
+
+	return compute.VirtualMachineScaleSetExtension{
+		Name: to.StringPtr(s.Name),
+		VirtualMachineScaleSetExtensionProperties: &compute.VirtualMachineScaleSetExtensionProperties{
+			Publisher:          to.StringPtr(s.Publisher),
+			Type:               to.StringPtr(s.Name),
+			TypeHandlerVersion: to.StringPtr(s.Version),
+			Settings:           nil,
+			ProtectedSettings:  s.ProtectedSettings,
+		},
+		// TODO: should we include location since it's used in VMExtensions too?
```

Comment:
not sure why one has location and not the other, is location required for VM extensions?

Reply:
The code in VMSS created the parameters w/o a location like this while the VM extensions included it. I kept the original behavior since it's a refactor but I wasn't sure if it was a bug.
```
extensions[i] = compute.VirtualMachineScaleSetExtension{
	Name: &extensionSpec.Name,
	VirtualMachineScaleSetExtensionProperties: &compute.VirtualMachineScaleSetExtensionProperties{
		Publisher:          to.StringPtr(extensionSpec.Publisher),
		Type:               to.StringPtr(extensionSpec.Name),
		TypeHandlerVersion: to.StringPtr(extensionSpec.Version),
		Settings:           nil,
		ProtectedSettings:  extensionSpec.ProtectedSettings,
	},
}
```

Reply:
I'm not sure either. You could try 1) removing Location from VM extensions 2) adding Location to VMSS extensions, and see if either breaks?

I agree it doesn't seem right that they aren't consistent, but it's possible that this is due to the Azure API between VM and VMSS not being consistent (if it's not, let's make it consistent).

Reply:
> Are VM/VMSS extensions covered in the e2e tests?

yes they are, since we create an extension for every single VM and every single VMSS to check kubeadm completion.

Reply:
from your other comment sounds like there is a discrepancy in the API and we can just get rid of this comment?